File: 58.cpp
#include <iostream>
#include <vector>
#include <stack>

using namespace std;

struct TreeNode {
    int val;
    struct TreeNode *left;
    struct TreeNode *right;
    TreeNode(int x) :
            val(x), left(NULL), right(NULL) {
    }
};

/*请实现一个函数，用来判断一颗二叉树是不是对称的。注意，如果一个二叉树同此二叉树的镜像是同样的，定义其为对称的。*/
/*思路：二叉树的镜像
比较中序遍历的正反，只适合每个结点保存不一样的值
故，将二叉树镜像之后，将每个结点进行比较*/

/*二叉树的镜像*/
TreeNode* tmp = NULL;
void mirrorTree(TreeNode* pRoot)
{
    if (pRoot == NULL)
        return;
    tmp = pRoot->left;
    pRoot->left = pRoot->right;
    pRoot->right = tmp;
    mirrorTree(pRoot->left);
    mirrorTree(pRoot->right);
}

/*二叉树的复制*/
TreeNode* copyTree(TreeNode* pRoot)
{
    if (pRoot == NULL)
        return NULL;
    TreeNode* newRoot = new TreeNode(pRoot->val);
    newRoot->left = copyTree(pRoot->left);
    newRoot->right = copyTree(pRoot->right);
    return newRoot;
}

/*比较镜像二叉树与原二叉树*/
bool compareTree(TreeNode* pRoot, TreeNode* newRoot)
{
    if (pRoot==NULL && newRoot==NULL)
        return true;
    if (pRoot!=NULL && newRoot==NULL)
        return false;
    if (pRoot==NULL && newRoot!=NULL)
        return false;
    if (pRoot->val != newRoot->val)
        return false;
    return compareTree(pRoot->left, newRoot->left) && compareTree(pRoot->right, newRoot->right);
}

bool isSymmetrical(TreeNode* pRoot)
{
    bool result = false;
    if (pRoot == NULL)
    {
        result = true;
        return result;
    }
    /*复制二叉树*/
    TreeNode* newRoot = copyTree(pRoot);
    /*将二叉树镜像*/
    mirrorTree(newRoot);
    /*比较镜像二叉树与原二叉树的每个结点*/
    result = compareTree(pRoot, newRoot);
    
    return result;
}

int main()
{
    return 0;
}

File: 58.py
# -*- coding:utf-8 -*-
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
class Solution:
    # 返回对应节点TreeNode
    def KthNode(self, pRoot, k):
        # write code here
        if not pRoot or k==0:
            return None
        result = []
        N = [0, k]
        self.MidSearch(pRoot, N, result)
        if k <= len(result):
            return result[k-1]
        else:
            return None
    
    def MidSearch(self, Root, N, result):
        if not Root or N[0] == N[1]:
            return
        self.MidSearch(Root.left, N, result)
        result.append(Root)
        N[0] += 1
        self.MidSearch(Root.right, N, result)

All = []
test = [8, 6, 10, 5, 7, 9, 11]
for n in test:
    All.append(TreeNode(n))
All[0].left = All[1]
All[0].right = All[2]
All[1].left = All[3]
All[1].right = All[4]
All[2].left = All[5]
All[2].right = All[6]
S = Solution()
print(S.KthNode(All[0], 0))